### PR TITLE
Added support for video analytics (and fix)

### PR DESCRIPTION
**Pull request by cat101**
- Added retry strategy for accounts with rate limit (HTTP 429)
- Changed VideoFile to disable sampling
- Fixed connectivity test issue on collab due to the lack of /etc/proto file  (and re-enabled connectivity test)

### CODE REVIEW

File: landingai/predict.py [hunk 1]
```diff
@@ -720,13 +726,13 @@ def _create_session(url: str, num_retry: int, headers: Dict[str, str]) -> Sessio
...
         total=num_retry,
-        backoff_factor=3,
+        backoff_factor=3,  # This the amount of seconds to wait on the second retry (i.e. 0, 3, 6, 12). First retry is immediate
         raise_on_redirect=True,
         raise_on_status=False,  # We are already raising exceptions during backend invocations
         allowed_methods=["GET", "POST", "PUT"],
         status_forcelist=[
             # 408 Request Timeout , 413 Content Too Large, 500 Internal Server Error
-            429,  # Too Many Requests  (ie. rate limiter)
+            # 429,  # Too Many Requests  (ie. rate limiter). This is handled externally
```
Comment (AsiaCao) [inline landingai/predict.py:735]: didn't quite get this part, how is the retry handled externally?
could you elaborate more about the line?

File: landingai/predict.py [hunk 2]
```diff
@@ -750,13 +756,24 @@ def _do_inference(
...
+        response = HttpResponse.from_response(resp)
+        if response.status_code != 429:
+            # The requests was successful or failed not due to the rate limiter
+            break
+        _LOGGER.warning(
+            "Got HTTP 429: You are exceeding your cloud inference rate. Sleeping for 60 sec"
         )
-    response = HttpResponse.from_response(resp)
+        time.sleep(60)
+
```
Comment (AsiaCao) [inline landingai/predict.py:776]: sense a little code smell with this "while true" loop. this logic combined with the existing Retry makes debugging and testing harder...   (is it worth it?)

can we tune the existing `backoff_factor` and other parameters in Retry instead? i.e. to let the Retry wait for a minute or half a minute

File: tests/unit/landingai/test_predict.py [hunk 3]
```diff
@@ -145,25 +144,6 @@ def test_predict_300():
     )
 
 
-@responses.activate
```
Comment (AsiaCao) [inline tests/unit/landingai/test_predict.py:148]: we should add an alternative test instead.

File: landingai/pipeline/image_source.py [hunk 4]
```diff
@@ -141,11 +141,36 @@ def __init__(self, uri: str, samples_per_second: float = 1) -> None:
...
+        cap.release()
+        if self._samples_per_second != 0:
+            # Adjust the video frames and fps based on sampling rate
+            self._target_total_frames = int(
+                self._src_total_frames * self._samples_per_second / self._src_fps
+            )
+            self._target_fps = self._samples_per_second
+        else:
+            self._target_total_frames = self._src_total_frames
+            self._target_fps = self._src_fps
```
Comment (AsiaCao) [inline landingai/pipeline/image_source.py:161]: could we add a unit test for it?

File: landingai/image_source_ops.py [hunk 5]
```diff
@@ -41,15 +41,18 @@ def sample_images_from_video(
...
     output_dir.mkdir(parents=True, exist_ok=True)
-    total_frames, sample_size, _ = probe_video(video_file, samples_per_second)
-    # Calculate the frame interval based on the desired frame rate
-    sample_interval = int(total_frames / sample_size)
+    if samples_per_second != 0:
+        total_frames, sample_size, _ = probe_video(video_file, samples_per_second)
+        # Calculate the frame interval based on the desired frame rate
+        sample_interval = int(total_frames / sample_size)
+    else:
+        sample_interval = 1
```
Comment (AsiaCao) [inline landingai/image_source_ops.py:55]: could we add a unit test to cover the path?

File: landingai/pipeline/image_source.py (hunk 4 above)
Comment (AsiaCao) [reply landingai/pipeline/image_source.py:161]: what's better is abstract out the `fps` resolution code as a dedicated function? so you don't have to test it twice (avoid duplication in `landingai/image_source_ops.py`)

File: landingai/predict.py (hunk 2 above)
Comment (cat101) [reply landingai/predict.py:776]: No. I tried to looking into urllib and request for tailored retry policies per error code...I even considered implementing a custom Retry class but it is not really designed to be extended.

About the `while True`, i do not like it but I could not think of an exit condition. It is ok for a customer that has 40 req/min to wait for a day to complete a video.... note that an exception like http 500 will break the look...

Comment (AsiaCao) [reply landingai/predict.py:776]: for a custom retry solution, you can explore the [Tenacity](https://tenacity.readthedocs.io/en/latest/) library if you feel like trying something new and have more time (our existing tests may need some tweaks).

File: landingai/image_source_ops.py (hunk 5 above)
Comment (cat101) [reply landingai/image_source_ops.py:55]: Done

File: tests/unit/landingai/test_predict.py (hunk 3 above)
Comment (AsiaCao) [reply tests/unit/landingai/test_predict.py:148]: yes, there is [a way](https://stackoverflow.com/questions/24897145/python-mock-multiple-return-values?noredirect=1&lq=1) to specify different behaviors for a series of calls. Though it's more work to fit into the current test.
maybe we can focus on solving [this one](https://github.com/landing-ai/landingai-python/pull/106#discussion_r1313404230) so we don't have to rework this test.

File: landingai/predict.py (hunk 2 above)
Comment (cat101) [reply landingai/predict.py]: Done. Used tenacity

File: tests/unit/landingai/test_predict.py (hunk 3 above)
Comment (cat101) [reply tests/unit/landingai/test_predict.py]: Done.. reverted back to original test

**Review by AsiaCao — APPROVED** [review]
nice work

**Comment by AsiaCao** [comment]
btw, don't forget to add a new entry to README to showcase your work, @cat101

**Pull request merged**